lib/membrane_http_adaptive_stream/hls_source/client_genserver.ex

```diff
@@ -6,46 +6,104 @@ defmodule Membrane.HLS.Source.ClientGenServer do
   use GenServer
   alias ExHLS.Client
 
-  @spec start_link(String.t(), ExHLS.DemuxingEngine.CMAF | ExHLS.DemuxingEngine.MPEGTS) ::
+  @spec start_link(
+          String.t(),
+          Membrane.HLS.Source.variant_selection_policy()
+        ) ::
           GenServer.on_start()
-  def start_link(url, demuxing_engine) do
-    GenServer.start_link(__MODULE__, url: url, demuxing_engine: demuxing_engine)
+  def start_link(url, variant_selection_policy) do
+    GenServer.start_link(__MODULE__,
+      url: url,
+      variant_selection_policy: variant_selection_policy
+    )
   end
 
-  @spec request_audio(pid()) :: :ok
-  def request_audio(client_genserver) do
-    GenServer.cast(client_genserver, {:request_audio, self()})
+  @spec request_audio_sample(pid()) :: :ok
+  def request_audio_sample(client_genserver) do
+    GenServer.cast(client_genserver, {:request_audio_sample, self()})
   end
 
-  @spec request_video(pid()) :: :ok
-  def request_video(client_genserver) do
-    GenServer.cast(client_genserver, {:request_video, self()})
+  @spec request_video_sample(pid()) :: :ok
+  def request_video_sample(client_genserver) do
+    GenServer.cast(client_genserver, {:request_video_sample, self()})
   end
 
+  # this function should be called by Membrane.HLS.Source
+  # before we start buffering the samples, to avoid waiting
+  # on downloading many segments
   @spec get_tracks_info(pid()) :: map()
   def get_tracks_info(client_genserver) do
     GenServer.call(client_genserver, :get_tracks_info)
   end
 
   @impl true
-  def init(url: url, demuxing_engine: demuxing_engine) do
-    client = Client.new(url, demuxing_engine)
-    variant_id = Client.get_variants(client) |> Enum.at(0) |> elem(0)
-    client = Client.choose_variant(client, variant_id)
-    {:ok, %{client: client}}
+  def init(url: url, variant_selection_policy: variant_selection_policy) do
+    state = %{
+      url: url,
+      variant_selection_policy: variant_selection_policy,
+      client: nil
+    }
+
+    # let's create Client and choose variant asnychronously
+    # beyond init/1, because it requires doing some HTTP requests
+    # so it can take some time
+    self() |> send(:setup)
```

Comment:
If it's just the matter of making sure that the parent of the GenServer doesn't need to wait for these HTTP requests, you can return `{:ok, state, {:continue, term()}` from the `init()` callback